src/main/java/com/google/collegeplanner/servlets/DatastoreServlet.java

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.collegeplanner.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.Entity;
+import com.google.collegeplanner.data.Course;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.http.client.utils.URIBuilder;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+
+/** Queries the UMD API and downloads the data to datastore. */
+@WebServlet("/api/download")
+public class DatastoreServlet extends BaseServlet {
+  DatastoreService datastore;
+
+  public DatastoreServlet() {
+    this(DatastoreServiceFactory.getDatastoreService(), new ApiUtil());
+  }
+
+  public DatastoreServlet(DatastoreService datastore, ApiUtil apiUtil) {
+    super(apiUtil);
+    this.datastore = datastore;
+  }
+
+  /*
+   * Downloads all course data to datastore.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Loop through the pages.
+    // We don't know how many pages there are beforehand. We stop looping when the new page
+    // doesn't have any results.
+    int page = 0;
+    do {
+      URI uri;
+      try {
+        URIBuilder builder = new URIBuilder("https://api.umd.io/v1/courses");
+        builder.setParameter("page", Integer.toString(page++));
+        uri = builder.build();
+      } catch (URISyntaxException e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+      JSONArray coursesArray = apiUtil.getJsonArray(uri);
+      if (coursesArray == null) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      } else if (coursesArray.size() == 0) {
+        respondWithError(HttpServletResponse.SC_OK, response);
+        return;
+      }
+      addCourses(coursesArray, response);
+    } while (true);
```

Comment:
instead of an actual infinite loop, I think you should set an upper limit (100 pages? 1000?) just in case. This should be declared as a constant.

This also helps protect against a case where the API returns the same thing no matter what page number you pass it.